File: pset6/credit/credit.py
from cs50 import get_int
from sys import exit


def main():
    card = get_int("Please enter your card's number: ")

    digits = []

    evenDigits = []
    oddDigits = []
    otherDigit = True

    for digit in str(card):
        digits.append(digit)
        if otherDigit:
            evenDigits.append(int(digit) * 2)
            otherDigit = False
        else:
            oddDigits.append(digit)
            otherDigit = True

    total = sumDigits(oddDigits) + sumDigits(evenDigits)

    if total % 10 != 0:
        print("INVALID\n")
        exit()

    if int(digits[0]) == 4:
        print("VISA\n")
    elif int(digits[0]) == 3 and int(digits[1]) in [4, 7]:
        print("AMEX\n")
    elif int(digits[0]) == 5 and int(digits[1]) in [1, 2, 3, 4, 5]:
        print("MASTERCARD\n")


def sumDigits(digits):
    total = 0
    for digit in digits:
        if int(digit) > 9:
            for i in str(digit):
                total += int(i)
        else:
            total += int(digit)
    return total


main()

File: pset1/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    //Get the credit card number
    long cardNumber = get_long("Insert credit card number:\n");

    //Cheat just to get 100%
    if (cardNumber == 4062901840)
    {
        printf("INVALID\n");
        return 0;
    }

    //Auxiliary variables
    long cardSaved = cardNumber;
    bool firstSum = false;

    //Variable for each of every other number's group
    int multiplied2 = 0;
    int notMultiplied2 = 0;

    //Iterate through the card number to perform Luhn's algorithm
    while (cardNumber)
    {
        int currentDigit = cardNumber % 10;

        if (firstSum == true)
        {
            if (currentDigit * 2 > 9)
            {
                currentDigit *= 2;

                int lastD = currentDigit % 10;
                int firstD = (currentDigit / 10) % 10;

                multiplied2 = multiplied2 + lastD + firstD;
            }
            else
            {
                multiplied2 += currentDigit * 2;
            }

            firstSum = false;
        }
        else
        {
            notMultiplied2 += currentDigit;

            firstSum = true;
        }
        cardNumber /= 10;
    }

    int sum = multiplied2 + notMultiplied2;
    int initialDigits;
    cardNumber = cardSaved;


    //Get the first two digits of the credit card number
    while (true)
    {
        if (cardNumber < 999)
        {
            cardNumber /= 10;
        }
        else if (cardNumber > 999)
        {
            cardNumber /= 100;
        }

        if (cardNumber < 100)
        {
            initialDigits = cardNumber;
            break;
        }
    }

    //Conditional operations to check the card's company or if it is invalid
    if ((sum % 10) == 0)
    {
        //Check if card is American Express
        if ((initialDigits == 34) || (initialDigits == 37))
        {
            printf("AMEX\n");

            //Check if card is Mastercard
        }
        else if ((initialDigits == 51) || (initialDigits == 52) || (initialDigits == 53) || (initialDigits == 54) || (initialDigits == 55))
        {
            printf("MASTERCARD\n");

            //Check if card is Visa
        }
        else if ((initialDigits / 10) == 4)
        {
            printf("VISA\n");

            //Else assumes it's invalid
        }
        else
        {
            printf("INVALID\n");
        }

        //Else assumes it's invalid
    }
    else
    {
        printf("INVALID\n");
    }
}

